pkg/registry/storage/csidriver/strategy_test.go

```diff
@@ -74,6 +78,222 @@ func TestCSIDriverStrategy(t *testing.T) {
 	}
 }
 
+func TestCSIDriverCreate(t *testing.T) {
+	ctx := genericapirequest.WithRequestInfo(genericapirequest.NewContext(), &genericapirequest.RequestInfo{
+		APIGroup:   "storage.k8s.io",
+		APIVersion: "v1beta1",
+		Resource:   "csidrivers",
+	})
+
+	attachRequired := true
+	podInfoOnMount := true
+	csiDriver := &storage.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+		},
+		Spec: storage.CSIDriverSpec{
+			AttachRequired: &attachRequired,
+			PodInfoOnMount: &podInfoOnMount,
+			VolumeModes: []storage.CSIVolumeMode{
+				storage.PersistentVolumeMode,
+			},
+		},
+	}
+
+	tests := []struct {
+		name       string
+		withInline bool
+	}{
+		{
+			name:       "inline enabled",
+			withInline: true,
+		},
+		{
+			name:       "inline disabled",
+			withInline: false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIInlineVolume, test.withInline)()
+
+			Strategy.PrepareForCreate(ctx, csiDriver)
+			errs := Strategy.Validate(ctx, csiDriver)
+			if len(errs) != 0 {
+				t.Errorf("unexpected validating errors: %v", errs)
+			}
+			if test.withInline {
+				if len(csiDriver.Spec.VolumeModes) != 1 {
+					t.Errorf("VolumeModes modified: %v", csiDriver.Spec)
+				}
+			} else {
+				if len(csiDriver.Spec.VolumeModes) != 0 {
+					t.Errorf("VolumeModes not stripped: %v", csiDriver.Spec)
+				}
+			}
+		})
+	}
+}
+
+func TestCSIDriverUpdate(t *testing.T) {
+	ctx := genericapirequest.WithRequestInfo(genericapirequest.NewContext(), &genericapirequest.RequestInfo{
+		APIGroup:   "storage.k8s.io",
+		APIVersion: "v1beta1",
+		Resource:   "csidrivers",
+	})
+
+	attachRequired := true
+	podInfoOnMount := true
+
+	tests := []struct {
+		name          string
+		withInline    bool
+		withDowngrade bool
+		old, update   *storage.CSIDriver
+		result        []storage.CSIVolumeMode
+	}{
+		{
+			name:       "add mode",
```

Comment:
To make sure we cover all the scenarios, it may be easier to structure the tests like:

- before: no mode, update: no mode
- before: no mode, update: add mode
- before: mode, update: change mode
- before: mode, update: drop mode

And run each with and without feature gate enabled.

Reply:
Done. The "downgrade" scenario is now covered by "before: mode, update: change mode" with feature disabled, because I changed the code so that it really only tests `PrepareForUpdate`, i.e. the feature gate doesn't influence the original object.